src/cljc/markdown/transformers.cljc

```diff
@@ -154,39 +153,51 @@
     [text state]))
 
 (defn close-paragraph [text {:keys [next-line paragraph] :as state}]
-  (if (and paragraph (= [\` \` \`] (take-last 3 (some-> next-line string/trim))))
+  (if (and paragraph (some-> next-line string/trim (string/ends-with? "```")))
     [(str text "</p>") (dissoc state :paragraph)]
     [text state]))
 
 (defn paragraph [text state]
   (apply close-paragraph (open-paragraph text state)))
 
-(defn code [text {:keys [eof lists code codeblock paragraph] :as state}]
-  (cond
-    (or lists codeblock)
-    [text state]
+(defn code [text {:keys [eof indent-code-end next-line lists code codeblock paragraph] :as state}]
+  (let [should-close? (or eof
+                          (not (or (string/blank? next-line)
+                                   (string/starts-with? next-line "    "))))]
+    (cond
+      (or lists codeblock)
+      [text state]
 
-    code
-    (if (or eof (not= "    " (string/join (take 4 text))))
-      [(str "</code></pre>" text) (dissoc state :indented-code :code :last-line-empty?)]
-      [(str "\n" (escape-code (string/replace-first text #"    " ""))) state])
+      indent-code-end
+      [text (-> state
+                (dissoc :code :indent-code-end :indented-code)
+                (assoc :last-line-empty? true))]
 
-    paragraph
-    [text state]
+      code
+      [(str (escape-code (string/replace-first text #"    " "\n"))
+            (when should-close? "</code></pre>"))
+       (cond-> state
+         should-close? (assoc  :indent-code-end true))]
 
-    (empty? (string/trim text))
-    [text state]
+      paragraph
+      [text state]
 
-    :default
-    (let [num-spaces (count (take-while (partial = \space) text))]
-      (if (> num-spaces 3)
-        [(str "<pre><code>" (escape-code (string/replace-first text #"    " "")))
-         (assoc state :code true :indented-code true)]
-        [text state]))))
+      (empty? (string/trim text))
+      [text state]
+
+      :default
+      (let [num-spaces (count (take-while (partial = \space) text))]
+        (if (>= num-spaces 4)
+          [(str "<pre><code>"
+                (escape-code (string/replace-first text #"    " ""))
+                (when should-close? "</code></pre>"))
+           (cond-> (assoc state :code true :indented-code true)
+             should-close? (assoc :indent-code-end true :last-line-empty? true))]
```

Comment:
- [ ] Recheck assumptions about having to add `last-line-empty` here

Reply:
yeah, I think that should be reasonable to do